nimbus/src/enrollment.rs

```diff
@@ -20,28 +24,367 @@ pub enum EnrolledReason {
+                        &self.slug
+                    );
+                    enrollment_events.push(EnrollmentChangeEvent::new(
+                        &self.slug,
+                        &enrollment_id,
+                        &branch,
+                        Some("optout"),
+                        EnrollmentChangeEventType::Disqualification,
+                    ));
+                    Self {
+                        slug: self.slug.clone(),
+                        status: EnrollmentStatus::Disqualified {
+                            reason: DisqualifiedReason::OptOut,
+                            enrollment_id,
+                            branch: branch.clone(),
+                        },
+                    }
+                } else {
+                    let updated_enrollment = evaluate_enrollment(
+                        nimbus_id,
+                        available_randomization_units,
+                        app_context,
+                        updated_experiment,
+                    )?;
+                    match updated_enrollment.status {
+                        EnrollmentStatus::Error { .. } => updated_enrollment,
+                        EnrollmentStatus::Enrolled {
+                            branch: ref new_branch,
+                            ..
+                        } if new_branch == branch => self.clone(),
+                        _ => {
+                            log::debug!(
+                            "Existing experiment enrollment '{}' is now disqualified (targeting change)",
+                            &self.slug
+                        );
+                            enrollment_events.push(EnrollmentChangeEvent::new(
+                                &self.slug,
+                                &enrollment_id,
+                                &branch,
+                                Some("targeting"),
+                                EnrollmentChangeEventType::Disqualification,
+                            ));
+                            Self {
+                                slug: self.slug.clone(),
+                                status: EnrollmentStatus::Disqualified {
+                                    reason: DisqualifiedReason::NotTargeted,
+                                    enrollment_id,
+                                    branch: branch.clone(),
+                                },
+                            }
+                        }
+                    }
+                }
+            }
+            EnrollmentStatus::Disqualified { .. } => {
+                if !is_user_participating {
+                    log::debug!(
+                        "Disqualified experiment enrollment '{}' has been reset to not-enrolled (global opt-out)",
+                        &self.slug
+                    );
+                    Self {
+                        slug: self.slug.clone(),
+                        status: EnrollmentStatus::NotEnrolled {
```

Comment:
They should stay in `Disqualified` but have `reason` set to `DisqualifiedReason::OptOut`, similar to what happens when we discover `!is_user_participating` while in the `Enrolled` state.  (Otherwise, the first transition after opting out would transition them from `Enrolled` to `Disqualified`, and the next one would take them from `Disqualified` to `NotEnrolled` and throw away their branch information).